gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -217,41 +238,72 @@ private void validate(GoogleCloudStorageItemInfo itemInfo) throws IOException {
-          "fetched metadata, resource:%s, time:%d", resourceId, (endTime - startTime));
-    }
-    return GoogleCloudStorageItemInfo.createObject(
-        resourceId,
-        /* creationTime= */ 0,
-        /* modificationTime= */ 0,
-        checkNotNull(object.getSize(), "size can not be null for '%s'", resourceId).longValue(),
-        /* contentType= */ null,
-        object.getContentEncoding(),
-        /* metadata= */ null,
-        checkNotNull(object.getGeneration(), "generation can not be null for '%s'", resourceId),
-        /* metaGeneration= */ 0,
-        /* verificationAttributes= */ null);
+    Sleeper sleeper = Sleeper.DEFAULT;
+    BackOff backoff = backOffFactory.newBackOff();
+    IOException exception = null;
+    do {
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      try {
+        Get metadataRequest = getMetadataRequest(gcs, resourceId).setFields(METADATA_FIELDS);
+        object = metadataRequest.execute();
+        publishSuccessMetric(LATENCY_MS, stopwatch, GET_OBJECT_METADATA, JSON);
+        return GoogleCloudStorageItemInfo.createObject(
+            resourceId,
+            /* creationTime= */ 0,
+            /* modificationTime= */ 0,
+            checkNotNull(object.getSize(), "size can not be null for '%s'", resourceId).longValue(),
+            /* contentType= */ null,
+            object.getContentEncoding(),
+            /* metadata= */ null,
+            checkNotNull(object.getGeneration(), "generation can not be null for '%s'", resourceId),
+            /* metaGeneration= */ 0,
+            /* verificationAttributes= */ null);
+      } catch (IOException ex) {
+        publishErrorMetric(LATENCY_MS, stopwatch, GET_OBJECT_METADATA, JSON, ex);
+        if (RetryDeterminer.SOCKET_ERRORS.shouldRetry(ex)) {
+          exception = ex;
+        } else {
+          throw errorExtractor.itemNotFound(ex)
+              ? createFileNotFoundException(resourceId, ex)
+              : new IOException("Error reading " + resourceId, ex);
+        }
+      } catch (Exception ex) {
+        publishErrorMetric(LATENCY_MS, stopwatch, GET_OBJECT_METADATA, JSON, ex);
+        throw ex;
+      }
+    } while (nextSleep(GET_OBJECT_METADATA, sleeper, backoff, exception));
+
+    throw errorExtractor.itemNotFound(exception)
+        ? createFileNotFoundException(resourceId, exception)
+        : new IOException("Error reading " + resourceId, exception);
+  }
+
+  private void publishSuccessMetric(
+      MeasureLong measure, Stopwatch stopwatch, String method, String protocol) {
+    long time = stopwatch.elapsed(MILLISECONDS);
+    TagKey[] keys = new TagKey[] {METHOD, STATUS, PROTOCOL};
+    String[] values = new String[] {method, OK, protocol};
+    metricsPublisher.recordLong(keys, values, measure, time);
+    logger.atFinest().log(
+        "method : %s , status : %s, protocol : %s , measure : %s , time : %d",
+        method, OK, protocol, measure, time);
+  }
+
+  private void publishErrorMetric(
```

Comment:
These methods are called `publish...` but they call `metricsPublisher.record...` methods. Are they actually publish anything or just record metrics that are published asynchronously?